xds/src/test/java/io/grpc/xds/ServerXdsClientNewServerApiTest.java

```diff
@@ -0,0 +1,690 @@
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.UInt32Value;
+import io.envoyproxy.envoy.config.core.v3.CidrRange;
+import io.envoyproxy.envoy.config.core.v3.SocketAddress;
+import io.envoyproxy.envoy.config.core.v3.TrafficDirection;
+import io.envoyproxy.envoy.config.core.v3.TransportSocket;
+import io.envoyproxy.envoy.config.listener.v3.Filter;
+import io.envoyproxy.envoy.config.listener.v3.FilterChain;
+import io.envoyproxy.envoy.config.listener.v3.FilterChainMatch;
+import io.envoyproxy.envoy.config.listener.v3.Listener;
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CommonTlsContext;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext;
+import io.envoyproxy.envoy.service.discovery.v3.AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceImplBase;
+import io.envoyproxy.envoy.service.discovery.v3.DiscoveryRequest;
+import io.envoyproxy.envoy.service.discovery.v3.DiscoveryResponse;
+import io.grpc.Context;
+import io.grpc.Context.CancellationListener;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.Status.Code;
+import io.grpc.SynchronizationContext;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.internal.FakeClock;
+import io.grpc.internal.FakeClock.TaskFilter;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.GrpcCleanupRule;
+import io.grpc.xds.EnvoyProtoData.Node;
+import io.grpc.xds.XdsClient.ListenerUpdate;
+import io.grpc.xds.XdsClient.ListenerWatcher;
+import io.grpc.xds.XdsClient.XdsChannel;
+import io.grpc.xds.XdsClientImpl2.ResourceType;
+import io.grpc.xds.internal.sds.CommonTlsContextTestsUtil;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/**
+ * Tests for {@link XdsClientImpl2 for server side Listeners using newServerApi}.
+ */
+@RunWith(JUnit4.class)
+public class ServerXdsClientNewServerApiTest {
```

Comment:
As mentioned elsewhere the class is set up with newServerApi=true in this case for all the tests. In the other file `ServerXdsClientTest` it is invoked with newServerApi=false and this whole file will eventually disappear when newServerApi=true is the only supported mode.

Reply:
This is a very verbose way though, with a lot of duplicates. Gating an unreleased feature with environment variable should be pure implementation detail and removing the environment variable should not require change how the class is used/created. Also, you shouldn't re-test logics not related to the feature the env variable is gating, you'd at most need one or two extra test cases verifying the xDS requests it sends are based on the environment value. Otherwise, if you need to change some of the behaviors unrelated to what the env variable is gating for, you will need to change the corresponding tests in both places.

Reply:
This is not different from the duplication of tests in `ServerXdsClientTest` and `XdsClientImplTestForListener`. In that case the tests are the same but the class being tested is different (`ServerXdsClient` and `XdsClientImpl2`). In this PR the class is the same but constructed with newServerApi=true vs false.

Also for every test the DiscoveryRequest is different (it has a resource name vs empty resource in the old test). Once the newServerApi becomes the norm the old code will be removed and any code related to `newServerApi==false` will be removed which is easier with the current structure. I prefer this structure to make it easy to just remove the old code.

Reply:
`XdsClientImplTestForListener` is intended to be deleted. It has been not because the client side is still using `XdsClientImpl2`. Once all usages of `XdsClientImpl2` are migrated, it will be deleted along with its tests (including `XdsClientImplTestForListener`).

Regarding duplicates (tests for stream retry) in `ClientXdsClientTest` and `ServerXdsClient`, they are run by completely different codepaths and developed by different authors, so it's acceptable to have two sets of tests for the purpose of easy collaboration. Ideally, we should have a way to cover stream retry logics only once.

But here, you are writing two complete sets of tests for the same piece of logic. You will need to maintain both sets together when there are new changes. Since this is your own code and no one else uses it, I am fine if you are ok with the maintenance burden or you feel this is easier for you.

Reply:
Believe me, this will make it easy for us to fall back to the old API in case the new API doesn't work with the control plane. And if it works it shouldn't take any time to remove the old code.